### PR TITLE
[fix] Add DeformableDetrFeatureExtractor

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

As pointed out by @Deppmeng who is adding Conditional DETR in #18948, the postprocessing of Deformable DETR is actually different compared to regular DETR. Namely, a sigmoid activation function is used rather than softmax, and the no-object class is included, whereas DETR discards this class.

Hence, we'll need a new `DeformableDetrFeatureExtractor` which includes this custom postprocessing logic. As only the postprocessing of object detection is different, I'm using `Copied from` statements wherever possible.

To do:

- [x] update `preprocessor_config.json` of all repos on the hub
- [x] use `from_pretrained` in the code snippets for the feature extractor

### CODE REVIEW

File: src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+            mask[: image.shape[1], : image.shape[2]] = True
+            pixel_mask.append(mask)
+
+        # return as BatchFeature
+        data = {"pixel_values": padded_images, "pixel_mask": pixel_mask}
+        encoded_inputs = BatchFeature(data=data, tensor_type=return_tensors)
+
+        return encoded_inputs
+
+    def post_process(self, outputs, target_sizes):
```
Comment (NielsRogge) [inline src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py:691]: FYI, this is the only difference with DETR's feature extractor.

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 2]
```diff
@@ -1599,14 +1599,14 @@ def forward(
         Examples:
 
         ```python
-        >>> from transformers import AutoFeatureExtractor, DeformableDetrModel
+        >>> from transformers import DeformableDetrFeatureExtractor, DeformableDetrModel
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:1602]: I'd leave the auto class here as we want to advertise them more, not less.

File: tests/models/deformable_detr/test_feature_extraction_deformable_detr.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        # create random PyTorch tensors
+        image_inputs = prepare_image_inputs(self.feature_extract_tester, equal_resolution=False, torchify=True)
+        for image in image_inputs:
+            self.assertIsInstance(image, torch.Tensor)
+
+        # Test whether the method "pad_and_return_pixel_mask" and calling the feature extractor return the same tensors
+        encoded_images_with_method = feature_extractor_1.pad_and_create_pixel_mask(image_inputs, return_tensors="pt")
+        encoded_images = feature_extractor_2(image_inputs, return_tensors="pt")
+
+        assert torch.allclose(encoded_images_with_method["pixel_values"], encoded_images["pixel_values"], atol=1e-4)
```
Comment (sgugger) [inline tests/models/deformable_detr/test_feature_extraction_deformable_detr.py:243]: Use `self.assertTrue` here and everywhere below for consistency.

File: src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+logger = logging.get_logger(__name__)
+
+
+ImageInput = Union[Image.Image, np.ndarray, "torch.Tensor", List[Image.Image], List[np.ndarray], List["torch.Tensor"]]
+
+
+def box_cxcywh_to_xyxy(x):
+    x_c, y_c, w, h = x.unbind(-1)
+    b = [(x_c - 0.5 * w), (y_c - 0.5 * h), (x_c + 0.5 * w), (y_c + 0.5 * h)]
+    return torch.stack(b, dim=-1)
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py:43]: This is the same as below, just with a more cryptic name -> to remove?

File: src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+        format (`str`, *optional*, defaults to `"coco_detection"`):
+            Data format of the annotations. One of "coco_detection" or "coco_panoptic".
+        do_resize (`bool`, *optional*, defaults to `True`):
+            Whether to resize the input to a certain `size`.
+        size (`int`, *optional*, defaults to 800):
+            Resize the input to the given size. Only has an effect if `do_resize` is set to `True`. If size is a
+            sequence like `(width, height)`, output size will be matched to this. If size is an int, smaller edge of
+            the image will be matched to this number. i.e, if `height > width`, then image will be rescaled to `(size *
+            height / width, size)`.
+        max_size (`int`, *optional*, defaults to `1333`):
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py:147]:
```suggestion
        max_size (`int`, *optional*, defaults to 1333):
```

File: src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py [hunk 6]
```diff
@@ -0,0 +1,952 @@
...
+    def post_process(self, outputs, target_sizes):
+        """
+        Converts the output of [`DeformableDetrForObjectDetection`] into the format expected by the COCO api. Only
+        supports PyTorch.
+
+        Args:
+            outputs ([`DeformableDetrObjectDetectionOutput`]):
+                Raw outputs of the model.
+            target_sizes (`torch.Tensor` of shape `(batch_size, 2)`):
+                Tensor containing the size (h, w) of each image of the batch. For evaluation, this must be the original
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py:700]:
```suggestion
                Tensor containing the size (height, width) of each image of the batch. For evaluation, this must be the original
```
Let's not use abbreviations in the doc!

**Review by sgugger — COMMENTED** [review]
Thanks for adding this new feature extractor! Left a couple of comments.

File: src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py (hunk 4 above)
Comment (NielsRogge) [reply src/transformers/models/deformable_detr/feature_extraction_deformable_detr.py:43]: Thanks for spotting that! Fixed

**Review by sgugger — APPROVED** [review]
You'll need to rebase on main to fix the tests, but otherwise LGTM!
